zebra-state/src/service/non_finalized_state/chain.rs

```diff
@@ -355,12 +355,42 @@ impl UpdateWith<Option<transaction::JoinSplitData<Groth16Proof>>> for Chain {
         joinsplit_data: &Option<transaction::JoinSplitData<Groth16Proof>>,
     ) -> Result<(), ValidateContextError> {
         if let Some(joinsplit_data) = joinsplit_data {
-            for sprout::JoinSplit { nullifiers, .. } in joinsplit_data.joinsplits() {
-                let span = debug_span!("revert_chain_state_with", ?nullifiers);
+            for nullifier in joinsplit_data.nullifiers() {
+                let span = debug_span!("update_chain_state_with", ?nullifier);
                 let _entered = span.enter();
-                trace!("Adding sprout nullifiers.");
-                self.sprout_nullifiers.insert(nullifiers[0]);
-                self.sprout_nullifiers.insert(nullifiers[1]);
+                trace!("adding sprout nullifier");
+
+                // Reject double-spends of nullifers:
+                // - both within the same JoinSplit,
+                // - from different JoinSplits in this transaction's JoinSplitData, or
+                // - one from this joinsplit, and another from:
+                //   - a previous transaction in this Block, or
+                //   - a previous block in this non-finalized Chain.
+                //
+                // (Duplicate finalized nullifiers are rejected during contextual validation.)
+                //
+                // TODO:
+                // - test that the chain's nullifiers are not modified on error
+                //
+                // "A nullifier MUST NOT repeat either within a transaction,
+                // or across transactions in a valid blockchain.
+                // Sprout and Sapling and Orchard nullifiers are considered disjoint,
+                // even if they have the same bit pattern."
+                //
+                // https://zips.z.cash/protocol/protocol.pdf#nullifierset
+                //
+                // "A transaction is not valid if it would have added a nullifier
+                // to the nullifier set that already exists in the set"
+                //
+                // https://zips.z.cash/protocol/protocol.pdf#commitmentsandnullifiers
+
+                // Reject the nullifier if it is already present in this non-finalized chain.
+                if !self.sprout_nullifiers.insert(*nullifier) {
+                    Err(ValidateContextError::DuplicateSproutNullifier {
+                        in_finalized_state: false,
+                        nullifier: *nullifier,
+                    })?;
+                }
```

Comment:
> But what if `update_chain_state_with` applies partial changes and returns an error?

I asked myself the same question, and made a PR to fix it:
https://github.com/ZcashFoundation/zebra/pull/2478

The previous code was functional (the entire chain was discarded), but potentially inefficient. It might also have done some weird things around chain forks, particularly if the previous blocks weren't available because the chain was dropped.

> I'm asking because I'm wondering if a malicious JoinSplit could be crafted that either:
>     * adds a valid nullifier to the state even though it wasn't mined yet, or

The nullifier is a hash of user-supplied values. I don't know the specifics, but I'm pretty sure the nullifier for a shielded output derived from the inputs (which must be unique) or a random value. So nullifiers are unique, as long as the hash remains computationally difficult to reverse.

>     * removes a valid nullifier from the state that was added by a previous valid transaction.

In Zebra, this isn't possible, because:
* all added nullifiers are checked to be unique
* blocks are only removed from a chain if they have previously been added to that chain